cerbero/build/source.py

```diff
@@ -408,14 +409,103 @@ class Git (GitCache):
     Source handler for git repositories
     '''
 
+    # Maps recipes with common sources share to the same fetch_done condition.
+    _fetch_done_cond_map = {}
+
     def __init__(self):
         GitCache.__init__(self)
         if self.commit is None:
             # Used by recipes in recipes/toolchain/
             self.commit = 'origin/sdk-%s' % self.version
+
+        self._fetching = False
         # For forced commits in the config
         self.commit = self.config.recipe_commit(self.name) or self.commit
 
+    def __getattribute__(self, attr_name):
+        attr = super().__getattribute__(attr_name)
+        if attr_name == "prepare":
+            self.__git_prepare = attr
+            raise AttributeError
+        return attr
+
+    def __getattr__(self, attr):
+        if attr == "prepare":
+            def _prepare():
+                # Set the fetch condition object after the default prepare
+                # method, as some recipes (properly) set/change remotes or
+                # repo_dir attributes from the prepare method, and the
+                # _update_fetch_done_cond method needs this info.
+                self.__git_prepare()
+                self._update_fetch_done_cond()
+            return _prepare
+        return super().__getattr__(attr)
+
+    def _get_source_id(self):
+        return (self.repo_dir, tuple(self.remotes.items()))
+
+    def _update_fetch_done_cond(self):
+        source_id = self._get_source_id()
+        if source_id in Git._fetch_done_cond_map:
+            return
+        lock = asyncio.Lock()
+        Git._fetch_done_cond_map[source_id] = asyncio.Condition(lock)
+
+    def _get_fetch_done_cond(self):
+        source_id = self._get_source_id()
+        return Git._fetch_done_cond_map[source_id]
+
+    def _get_git_recipes(self):
+        recipes = self.config.cookbook.get_recipes_list()
+        return [r for r in recipes if issubclass(r.stype, SourceType.GIT)]
+
+    def _common_source_recipes(self):
+        target_source_id = self._get_source_id()
+        return [r for r in self._get_git_recipes() if target_source_id == r._get_source_id()]
+
+    def _already_fetched(self):
+        for recipe in self._common_source_recipes():
+            status = self.config.cookbook.status.get(recipe.name)
+            if status and 'fetch' in status.steps:
+                return True
+        return False
+
+    def _other_fetching(self):
+        for recipe in self._common_source_recipes():
+            if recipe == self:
+                continue
+            if recipe._fetching:
+                return True
+        return False
+
+    async def fetch(self, **kwargs):
+        self._fetching = True
+
+        fetch_done = self._get_fetch_done_cond()
+        already_fetched = self._already_fetched()
```

Comment:
We're calling `_common_source_recipes` in both methods, so I think as an idea we could call it once and then pass it as a parameter to `_already_fetched` and `_other_fetching`. Considering it iterates through all recipes, I think it's worth it.